File: tests/mqtt/client/paho-jwt-auth.py
'''
    Silverline Client MQTT JWT authentication example for Python paho-mqtt
'''

import ssl

import paho.mqtt.client as mqtt


class Silverline:

    def connect_client_test(self, host, username, token, verify=True):
        port = 8883
        client = mqtt.Client()
        client.username_pw_set(username=username, password=token)

        if verify(host):
            # CA-signed
            client.tls_set()
        else:
            # self-signed
            client.tls_set_context(ssl._create_unverified_context())
            client.tls_insecure_set(True)
        try:
            client.connect(host, port=port)
        except Exception as err:
            print(f'MQTT connect error to {host}, port={port}: {err}')

        client.disconnect()

File: tests/mqtt/client/paho-jwt-auth.c
/**
 *  Silverline Client MQTT JWT authentication example for Paho MQTT C
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "MQTTClient.h"

#define CLIENTID "ExampleClientSilverline"

int client_connect(char *host, char *token, char *password, bool verify)
{
    MQTTClient client;
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    MQTTClient_SSLOptions ssl_opts = MQTTClient_SSLOptions_initializer;
    int rc;
    MQTTClient_create(&client, &host, CLIENTID,
                      MQTTCLIENT_PERSISTENCE_NONE, NULL);

    // verify ? CA-signed : self-signed
    ssl_opts.verify = verify;
    ssl.opts.enableServerCertAuth = verify;
    ssl_opts.sslVersion = MQTT_SSL_VERSION_TLS_1_2;

    conn_opts.ssl = ssl_opts;
    conn_opts.cleansession = 1;
    conn_opts.username = &token;
    conn_opts.password = &password;

    if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS)
    {
        printf("Failed to connect, return code %d\n", rc);
        exit(EXIT_FAILURE);
    }
    MQTTClient_disconnect(client, 10000);
    MQTTClient_destroy(&client);
    return rc;
}
